lints/lint_rsa_algid_null_params.go

```diff
@@ -0,0 +1,103 @@
+/*
+ * ZLint Copyright 2019 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package lints
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"fmt"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+type rsaAlgIDNullParams struct{}
+
+// Initialize is a NOP.
+func (l *rsaAlgIDNullParams) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true for certificates with an RSA public key.
+func (l *rsaAlgIDNullParams) CheckApplies(c *x509.Certificate) bool {
+	_, ok := c.PublicKey.(*rsa.PublicKey)
+	return ok && c.PublicKeyAlgorithm == x509.RSA
+}
+
+// Execute returns a fatal lint result if the provided certificate encodes the
+// RSA algorithm identifier SEQUENCE without the required NULL parameters.
+//
+// The baseline requirements mandate that certificates obey RFC 5280, which in
+// turn defers to RFC 3279 for the encoding of the RSA algorithm identifier. RFC
+// 3279 Section 2.3.1 "RSA keys" says:
+//
+//   The rsaEncryption OID is intended to be used in the algorithm field
+//   of a value of type AlgorithmIdentifier.  The parameters field MUST
+//   have ASN.1 type NULL for this algorithm identifier.
+//
+// This lint verifies that the NULL parameters MUST of this section is met.
+func (l *rsaAlgIDNullParams) Execute(c *x509.Certificate) *LintResult {
+	raw := c.SubjectAndKey().RawSubjectPublicKeyInfo
+
+	var pkixPublicKey struct {
+		Algo      pkix.AlgorithmIdentifier
+		BitString asn1.BitString
+	}
+
+	if rest, err := asn1.Unmarshal(raw, &pkixPublicKey); err != nil {
+		return &LintResult{
+			Status:  Error,
+			Details: fmt.Sprintf("Error unmarshaling raw pkixPublicKey: %v", err),
+		}
+	} else if len(rest) != 0 {
+		return &LintResult{
+			Status:  Error,
+			Details: fmt.Sprintf("Trailing data after pkixPublicKey: %v", err),
+		}
+	}
+
+	params := pkixPublicKey.Algo.Parameters
+
+	if params.Tag != asn1.TagNull {
```

Comment:
Looking at how asn1.unmarshal is implemented, doesn't this create the risk that either the class or the compound-type may be misconstructed? The memcmp approach proposed from the associated thread was, in part, to ensure that it'd be flagged as a Universal class, non-compound. If you want to provide more descriptiveness, then it's probably worth adding checks.

One way you could restructure this to be more defense-in-depth would be to early execute the compare with the 'raw' value. If the raw value was equal, you know it's a pass, otherwise, you know there is 'something' wrong. The unmarshal and the remainder of the function is thus trying to find out 'what' is wrong and add a descriptive error to it.

WDYT of that approach?

Reply:
That sounds more robust to me. I'll take a shot at doing that over the next few days.

Reply:
Actually knocked up this lint using cryptobyte today before I saw this PR.
Sharing here in case you wanted a look.

```
        var input = cryptobyte.String(c.RawSubjectPublicKeyInfo)

	var publicKeyInfo cryptobyte.String
	if !input.ReadASN1(&publicKeyInfo, asn1_cryptobyte.SEQUENCE) {
		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey"}
	}

	var algorithm cryptobyte.String
	if !publicKeyInfo.ReadASN1(&algorithm, asn1_cryptobyte.SEQUENCE) {
		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey algorithm"}
	}

	if !algorithm.SkipASN1(asn1_cryptobyte.OBJECT_IDENTIFIER) {
		return &LintResult{Status: Fatal, Details: "error reading pkixPublicKey algorithm OID"}
	}

	if algorithm.Empty() {
		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier missing required NULL parameter"}
	}

	if !algorithm.PeekASN1Tag(asn1_cryptobyte.NULL) {
		return &LintResult{Status: Error, Details: "certificate contains RSA public key algorithm identifier with non-NULL parameter"}
	}

	return &LintResult{Status: Pass}
```